File: JAVA/2641. Cousins in Binary Tree II.java
class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int levelSum=root.val;
        while(!q.isEmpty())
        {
            int n = q.size();
            int nextLevelSum = 0;
            for(int i=0;i<n;i++)
            {
                TreeNode curr = q.poll();
                curr.val = levelSum-curr.val;

                int siblingSum = curr.left!=null ? curr.left.val : 0;
                siblingSum += curr.right!=null ? curr.right.val : 0;

                if(curr.left!=null)
                {
                    nextLevelSum += curr.left.val;
                    curr.left.val = siblingSum;
                    q.add(curr.left);
                }
                 if(curr.right!=null)
                {
                    nextLevelSum += curr.right.val;
                    curr.right.val = siblingSum;
                    q.add(curr.right);
                }
            }
            levelSum = nextLevelSum;
        }
        return root;
    }
}

File: C++/2641. Cousins in Binary Tree II.cpp
class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        if(root == NULL)
            return {};
        if(root->left == NULL && root->right == NULL){
            root->val = 0;
            return root;
        }
        queue<TreeNode*> q;
        q.push(root);
        root->val = 0;
        unordered_map<TreeNode*, TreeNode*> m; 
        unordered_map<TreeNode*, int> m2; 
        while(!q.empty()){
            int n = q.size();
            m.clear();
            int sum = 0;
            while(n--){
                TreeNode* curr = q.front();
                q.pop();
                m2[curr] = 0;
                if(curr->left != NULL){
                    q.push(curr->left);
                    m[curr->left] = curr;
                    sum += curr->left->val;
                    m2[curr] += curr->left->val;
                }  
                if(curr->right != NULL){
                    q.push(curr->right);
                    m[curr->right] = curr;
                    sum += curr->right->val;
                    m2[curr] += curr->right->val;
                }  

            }
            for(auto& it : m){
                TreeNode* child = it.first;
                TreeNode* parent = it.second;
                child->val = sum - m2[parent];
            }
        }
        return root;        
    }
};
